lib/promscrape/scrapework_timing_test.go

```diff
@@ -96,3 +99,97 @@ vm_tcplistener_write_calls_total{name="https", addr=":443"} 132356
+vm_tcplistener_conns{name="http", addr=":80"} 0
+vm_tcplistener_conns{name="https", addr=":443"} 2
+vm_tcplistener_errors_total{name="http", addr=":80", type="accept"} 0
+vm_tcplistener_errors_total{name="http", addr=":80", type="close"} 0
+vm_tcplistener_errors_total{name="http", addr=":80", type="read"} 97
+vm_tcplistener_errors_total{name="http", addr=":80", type="write"} 2
+vm_tcplistener_errors_total{name="https", addr=":443", type="accept"} 0
+vm_tcplistener_errors_total{name="https", addr=":443", type="close"} 0
+vm_tcplistener_errors_total{name="https", addr=":443", type="read"} 243
+vm_tcplistener_errors_total{name="https", addr=":443", type="write"} 285
+vm_tcplistener_read_bytes_total{name="http", addr=":80"} 879339
+vm_tcplistener_read_bytes_total{name="https", addr=":443"} 19453340
+vm_tcplistener_read_calls_total{name="http", addr=":80"} 7780
+vm_tcplistener_read_calls_total{name="https", addr=":443"} 70323
+vm_tcplistener_read_timeouts_total{name="http", addr=":80"} 673
+vm_tcplistener_read_timeouts_total{name="https", addr=":443"} 12353
+vm_tcplistener_write_calls_total{name="http", addr=":80"} 3996
+vm_tcplistener_write_calls_total{name="https", addr=":443"} 132356
+`
+	protoparserutil.StartUnmarshalWorkers()
+	defer protoparserutil.StopUnmarshalWorkers()
+
+	readDataFunc := func(dst *bytesutil.ByteBuffer) error {
+		dst.B = append(dst.B, data...)
+		return nil
+	}
+	b.ReportAllocs()
+	b.SetBytes(int64(len(data)))
+	b.RunParallel(func(pb *testing.PB) {
+		var sw scrapeWork
+		sw.Config = &ScrapeWork{
+			StreamParse: true,
+		}
+		sw.ReadData = readDataFunc
+		sw.PushData = func(_ *auth.Token, _ *prompbmarshal.WriteRequest) {}
+		tsmGlobal.Register(&sw)
+		timestamp := int64(0)
+		for pb.Next() {
+			if err := sw.scrapeInternal(timestamp, timestamp); err != nil {
+				panic(fmt.Errorf("unexpected error: %w", err))
+			}
+			timestamp++
+		}
+		tsmGlobal.Unregister(&sw)
+	})
+}
+
+func BenchmarkScrapeWorkScrapeInternalStreamBigData(b *testing.B) {
+	generateScrape := func(n int) string {
+		w := strings.Builder{}
+		for i := 0; i < n; i++ {
+			w.WriteString(fmt.Sprintf("fooooo_%d 1\n", i))
+		}
+		return w.String()
+	}
+
+	data := generateScrape(200000)
+	protoparserutil.StartUnmarshalWorkers()
+	defer protoparserutil.StopUnmarshalWorkers()
+
+	readDataFunc := func(dst *bytesutil.ByteBuffer) error {
+		dst.B = append(dst.B, data...)
+		return nil
+	}
+
+	var sw scrapeWork
+	sw.Config = &ScrapeWork{
+		StreamParse: true,
+	}
+	sw.ReadData = readDataFunc
+	sw.PushData = func(_ *auth.Token, _ *prompbmarshal.WriteRequest) {
+		// simulates a delay to highlight the difference between lock-based and lock-free algorithms.
+		// See https://github.com/VictoriaMetrics/VictoriaMetrics/pull/8515
+		time.Sleep(time.Millisecond)
+	}
+	tsmGlobal.Register(&sw)
+	defer tsmGlobal.Unregister(&sw)
+
+	b.ReportAllocs()
+	b.SetBytes(int64(len(data)))
+
+	timestamp := int64(0)
+	for b.Loop() {
```

Comment:
When running the benchmark, I noticed that 90% of time is spent on reading the body rather than on unmarshalling it. So relatively, it measures the performace of `stream.Parse` calls rather than `processDataInStreamMode` performance. 
However, it indeed proves that `processDataInStreamMode` has no impact anymore. That's cool! 
But for further improvements, should we re-consider concurrent read calls? So far it reads the body in single thread and unmarshals it with gomaxprocs threads.